### PR TITLE
#54 - Added Tag class and it's usage in ManifestRef

### PR DESCRIPTION
**Pull request by olegmoz**
Part of issue #54 
Tag is one of important concepts in Docker API and it's validation is required to implement HTTP API properly.

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#63](https://github.com/artipie/docker-adapter/issues/63) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/dc0c4ca4-fefd-4b50-af1d-21cb379082f8); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by olegmoz** [comment]
@Vatavuk ping

File: src/test/java/com/artipie/docker/TagValidTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        "1.0",
+        "my-tag",
+        "MY_TAG",
+        "My.Tag.1",
+        "_some_tag",
+        //@checkstyle LineLengthCheck (1 line)
+        "01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567"
+    })
+    void shouldGetValueWhenValid(final String original) {
+        final Tag tag = new Tag.Valid(original);
```
Comment (Vatavuk) [inline src/test/java/com/artipie/docker/TagValidTest.java]: @olegmoz let's inline this variable

File: src/test/java/com/artipie/docker/TagValidTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    @ValueSource(strings = {
+        "",
+        "*",
+        "\u00ea",
+        "-my-tag",
+        //@checkstyle LineLengthCheck (1 line)
+        "012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678"
+    })
+    void shouldFailToGetValueWhenInvalid(final String original) {
+        final Tag tag = new Tag.Valid(original);
```
Comment (Vatavuk) [inline src/test/java/com/artipie/docker/TagValidTest.java]: @olegmoz let's inline this variable

File: src/test/java/com/artipie/docker/TagValidTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        "01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567"
+    })
+    void shouldGetValueWhenValid(final String original) {
+        final Tag tag = new Tag.Valid(original);
+        MatcherAssert.assertThat(tag.value(), new IsEqual<>(original));
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {
+        "",
```
Comment (Vatavuk) [inline src/test/java/com/artipie/docker/TagValidTest.java]: @olegmoz let's add one more variant `.0`

File: src/main/java/com/artipie/docker/Tag.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+package com.artipie.docker;
+
+import java.util.regex.Pattern;
+
+/**
+ * Docker image tag.
+ * See <a href="https://docs.docker.com/engine/reference/commandline/tag/">docker tag</a>.
+ *
+ * @since 0.1
```
Comment (Vatavuk) [inline src/main/java/com/artipie/docker/Tag.java]: @olegmoz `0.2`

**Comment by olegmoz** [comment]
@Vatavuk thanks, made changes by your comments

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 11@4ac5e2bff6a6, thread: PQ-CT2E6TK9B
com.zerocracy.farm.ruled.RdRules[130] java.lang.IllegalStateException: http://datum.zerocracy.com/latest/rules/pm/no-lost-estimates.xsl says: The estimate for job "gh:artipie/docker-adapter#63" exists in estimates.xml, but the job is absent in the WBS among 33 jobs.
```

[0.52.11](https://github.com/zerocracy/farm): CID: [1a3f5bff-db72-4bfd-9387-a1b50c9bb40b](https://www.0crat.com/footprint/CT2E6TK9B/1a3f5bff-db72-4bfd-9387-a1b50c9bb40b), Type: "Remove job from WBS"

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @vatavuk[/z](https://www.0crat.com/u/vatavuk)[/z](https://www.0crat.com/u/vatavuk) didn't make CR comments

**Pull request merged**